### PR TITLE
Generate Appwrite ID's on the Client-Side

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?

This PR Implements Appwrite's ID generation onto the Client/Server side SDKs

## Test Plan
Existing tests have been modified to check these ID's are being generated and do not match `unique()` anymore.

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

File: templates/android/library/src/main/java/io/appwrite/ID.kt.twig [hunk 1]
```diff
@@ -1,10 +1,30 @@
 package {{ sdk.namespace | caseDot }}
 
+import java.time.Instant
+import kotlin.math.floor
+import kotlin.random.Random
+
+fun uniqid(): String {
```
[the remark this answers is not in the document]
Comment (gewenyu99) [reply templates/android/library/src/main/java/io/appwrite/ID.kt.twig:7]: This function just generates a timestamp so our IDs are in order, no?

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: templates/dart/lib/id.dart.twig [hunk 2]
```diff
@@ -4,13 +4,35 @@ part of {{ language.params.packageName }};
 class ID {
   ID._();
 
-  /// Have Appwrite generate a unique ID for you.
-  static String unique() {
-    return 'unique()';
+  // Generate a unique ID based on timestamp
```
[the remark this answers is not in the document]
Comment (PineappleIOnic) [reply templates/dart/lib/id.dart.twig:7]: Sure

File: templates/ruby/lib/container/id.rb.twig [hunk 3]
```diff
@@ -1,11 +1,28 @@
+require 'securerandom'
+
+#Generate a unique ID based on timestamp
+#Recreated from https://www.php.net/manual/en/function.uniqid.php
+def uniqueId
```
Comment (gewenyu99) [inline templates/ruby/lib/container/id.rb.twig:5]: use snake case

File: templates/android/library/src/main/java/io/appwrite/ID.kt.twig [hunk 4]
```diff
@@ -1,10 +1,35 @@
 package {{ sdk.namespace | caseDot }}
 
+import java.time.Instant
+import kotlin.math.floor
+import kotlin.random.Random
+
+// Generate a unique ID based on timestamp
+// Recreated from https://www.php.net/manual/en/function.uniqid.php
+private fun uniqueId(): String {
```
Comment (gewenyu99) [inline templates/android/library/src/main/java/io/appwrite/ID.kt.twig:9]: should be called hex timestamp

File: templates/android/library/src/main/java/io/appwrite/ID.kt.twig [hunk 5]
```diff
@@ -1,10 +1,35 @@
 package {{ sdk.namespace | caseDot }}
 
+import java.time.Instant
+import kotlin.math.floor
+import kotlin.random.Random
+
+// Generate a unique ID based on timestamp
```
Comment (gewenyu99) [inline templates/android/library/src/main/java/io/appwrite/ID.kt.twig:7]: This is wrong.

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: templates/node/lib/id.js.twig [hunk 6]
```diff
@@ -1,7 +1,29 @@
...
+    static #hexTimestamp = () => {
+        const now = new Date();
+        const secondsSinceEpoch = Math.floor(now.getTime() / 1000);
+        const msecs = now.getTime() - secondsSinceEpoch * 1000;
+        const microsecondsPart = msecs * 1000; // Convert milliseconds to microseconds
 
-    static unique = () => {
-         return 'unique()'
+        // Convert to hexadecimal
+        const hexTimestamp = secondsSinceEpoch.toString(16) + microsecondsPart.toString(16).padStart(5, '0');
```
Comment (gewenyu99) [inline templates/node/lib/id.js.twig:11]: This line is too long.

File: templates/web/src/id.ts.twig [hunk 7]
```diff
@@ -1,9 +1,27 @@
+// Generate an hex ID based on timestamp
+// Recreated from https://www.php.net/manual/en/function.uniqid.php
+function hexTimestamp(): string {
```
Comment (gewenyu99) [inline templates/web/src/id.ts.twig:3]: Move this into the class as a private.

File: templates/web/src/id.ts.twig [hunk 8]
```diff
@@ -1,9 +1,23 @@
+function uniqid(): string {
+  const now = new Date();
+  const secondsSinceEpoch = Math.floor(now.getTime() / 1000);
+  const microsecondsPart = now.getMilliseconds() * 1000; // Convert milliseconds to microseconds
+  const hexTimestamp = `${secondsSinceEpoch.toString(16)}${microsecondsPart.toString(16).padStart(5, '0')}`;
```
[the remark this answers is not in the document]
Comment (gewenyu99) [reply templates/web/src/id.ts.twig:7]: Break this up into multiple lines.

File: templates/swift/Sources/ID.swift.twig [hunk 9]
```diff
@@ -1,9 +1,27 @@
+import Foundation
+
+// Generate an hex ID based on timestamp
+// Recreated from https://www.php.net/manual/en/function.uniqid.php
+func hexTimestamp() -> String {
```
Comment (gewenyu99) [inline templates/swift/Sources/ID.swift.twig:5]: Make this private and move it in.

File: templates/swift/Sources/ID.swift.twig [hunk 10]
```diff
@@ -1,9 +1,27 @@
...
+    let now = Date()
+    let secondsSinceEpoch = Int(now.timeIntervalSince1970)
+    let microsecondsPart = Int((now.timeIntervalSince1970 - Double(secondsSinceEpoch)) * 1_000_000)
```
Comment (gewenyu99) [inline templates/swift/Sources/ID.swift.twig:8]: Cna this reference the secondsSinceEpoch part? Not sure. 

secondsSinceEpoch - Double(secondsSinceEpoch)


It's weird that you call a `now.timeIntervalSince1970` twice

File: templates/ruby/lib/container/id.rb.twig [hunk 11]
```diff
@@ -1,11 +1,28 @@
+require 'securerandom'
+
+#Generate an hex ID based on timestamp
+#Recreated from https://www.php.net/manual/en/function.uniqid.php
+def hex_timestamp
```
Comment (gewenyu99) [inline templates/ruby/lib/container/id.rb.twig:5]: Private, move this into the class.

File: templates/python/package/id.py.twig [hunk 12]
```diff
@@ -1,8 +1,25 @@
...
 
+    # Generate a unique ID with padding to have a longer ID
+    # Recreated from https://github.com/utopia-php/database/blob/main/src/Database/ID.php#L13
     @staticmethod
-    def unique():
-        return 'unique()'
\ No newline at end of file
+    def unique(padding = 7):
+        base_id = ID.__hex_timestamp()
+        random_padding = ''.join(random.choice('0123456789abcdef') for _ in range(padding))
```
Comment (gewenyu99) [inline templates/python/package/id.py.twig:24]: Use 
```py
random_bytes = os.urandom(6)
hex_string = random_bytes.hex()` or something.
```

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: templates/deno/src/id.ts.twig [hunk 13]
```diff
@@ -1,9 +1,31 @@
 export class ID {
+    // Generate an hex ID based on timestamp
+    // Recreated from https://www.php.net/manual/en/function.uniqid.php
+    static #hexTimestamp(): string {
+        const now = new Date();
+        const secondsSinceEpoch = Math.floor(now.getTime() / 1000);
+        const msecs = now.getTime() - secondsSinceEpoch * 1000;
+        const microsecondsPart = msecs * 1000; // Convert milliseconds to microseconds
```
Comment (gewenyu99) [inline templates/deno/src/id.ts.twig:8]: Why don't we go to this directly like you do in: `    final msecs = now.microsecondsSinceEpoch - secondsSinceEpoch * 1000000;`

File: templates/dotnet/src/Appwrite/ID.cs.twig [hunk 14]
```diff
@@ -1,15 +1,44 @@
...
+        // Generate an hex ID based on timestamp
+        // Recreated from https://www.php.net/manual/en/function.uniqid.php
+        private static string HexTimestamp()
         {
-            return "unique()";
+            var now = DateTime.UtcNow;
+            var epoch = (now - new DateTime(1970, 1, 1));
+            var secondsSinceEpoch = (long)epoch.TotalSeconds;
+            var msecs = epoch.TotalMilliseconds - secondsSinceEpoch * 1000;
+            var microsecondsPart = (long)(msecs * 1000); // Convert milliseconds to microseconds
```
Comment (gewenyu99) [inline templates/dotnet/src/Appwrite/ID.cs.twig:15]: Why don't you go directly like:     final msecs = now.microsecondsSinceEpoch - secondsSinceEpoch * 1000000;

File: templates/ruby/lib/container/id.rb.twig [hunk 15]
```diff
@@ -1,11 +1,29 @@
...
+            random_padding = SecureRandom.hex(padding)
+            random_padding = random_padding[0...padding]
+            base_id + random_padding
+        end
+
+        private
+
+        #Generate an hex ID based on timestamp
+        #Recreated from https://www.php.net/manual/en/function.uniqid.php
+        def self.hex_timestamp
```
Comment (gewenyu99) [inline templates/ruby/lib/container/id.rb.twig:21]: > You can use private_class_method :method_name if you want to define a class method as private

File: templates/web/src/id.ts.twig [hunk 16]
```diff
@@ -1,9 +1,27 @@
 export class ID {
+    // Generate an hex ID based on timestamp
+    // Recreated from https://www.php.net/manual/en/function.uniqid.php
+    static #hexTimestamp(): string {
+        const now = new Date();
+        const secondsSinceEpoch = Math.floor(now.getTime() / 1000); // Get Unix Timestamp
+        const microsecondsPart = now.getMilliseconds() * 1000; // Convert milliseconds to microseconds
```
Comment (gewenyu99) [inline templates/web/src/id.ts.twig:7]: Should this not be 
```suggestion
        const microsecondsPart = now.getMilliseconds() * 1000000; // Convert milliseconds to microseconds
```

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: templates/android/library/src/main/java/io/package/ID.kt.twig [hunk 17]
```diff
@@ -1,10 +1,34 @@
 package {{ sdk.namespace | caseDot }}
 
+import java.time.Instant
```
Comment (gewenyu99) [inline templates/android/library/src/main/java/io/package/ID.kt.twig:3]: This works good

File: templates/dart/lib/id.dart.twig [hunk 18]
```diff
@@ -4,13 +4,34 @@ part of {{ language.params.packageName }};
 class ID {
   ID._();
 
-  /// Have Appwrite generate a unique ID for you.
-  static String unique() {
-    return 'unique()';
+  // Generate an hex ID based on timestamp
+  // Recreated from https://www.php.net/manual/en/function.uniqid.php
+  static String _hexTimestamp() {
```
Comment (gewenyu99) [inline templates/dart/lib/id.dart.twig:9]: Seems to work, I think the timestamp behavior here is a little different from Kotlin.

File: templates/deno/src/id.ts.twig [hunk 19]
```diff
@@ -1,9 +1,30 @@
 export class ID {
+    // Generate an hex ID based on timestamp
+    // Recreated from https://www.php.net/manual/en/function.uniqid.php
+    static #hexTimestamp(): string {
+        const now = new Date();
```
Comment (gewenyu99) [inline templates/deno/src/id.ts.twig:5]: Seems to work, but some of the IDs aren't ordered, the timestamp bit is the same.

File: templates/dotnet/src/Appwrite/ID.cs.twig [hunk 20]
```diff
@@ -1,15 +1,42 @@
+using System;
+
 namespace {{ spec.title | caseUcfirst }}
```
Comment (gewenyu99) [inline templates/dotnet/src/Appwrite/ID.cs.twig:3]: Works reliably

File: templates/kotlin/src/main/kotlin/io/appwrite/ID.kt.twig [hunk 21]
```diff
@@ -1,10 +1,33 @@
 package {{ sdk.namespace | caseDot }}
```
Comment (gewenyu99) [inline templates/kotlin/src/main/kotlin/io/appwrite/ID.kt.twig:1]: Good

File: templates/node/lib/id.js.twig [hunk 22]
```diff
@@ -1,7 +1,27 @@
 class ID {
```
Comment (gewenyu99) [inline templates/node/lib/id.js.twig:1]: Has the padding issues.

File: templates/php/src/ID.php.twig [hunk 23]
```diff
@@ -3,13 +3,20 @@
 namespace {{ spec.title | caseUcfirst }};
```
Comment (gewenyu99) [inline templates/php/src/ID.php.twig:3]: Perfect

File: templates/python/package/id.py.twig [hunk 24]
```diff
@@ -1,8 +1,26 @@
+from datetime import datetime
```
Comment (gewenyu99) [inline templates/python/package/id.py.twig:1]: Padding issues.

File: templates/ruby/lib/container/id.rb.twig [hunk 25]
```diff
@@ -1,11 +1,27 @@
+require 'securerandom'
```
Comment (gewenyu99) [inline templates/ruby/lib/container/id.rb.twig:1]: Works well

**Review by gewenyu99 — APPROVED** [review]

**Review by eldadfux — APPROVED** [review]

**Pull request merged**